server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java

```diff
@@ -19,39 +19,208 @@
+            Optional<String> body = content.getTextBody();
+            if (body.isPresent()) {
+                return body;
+            } else {
+                return content.getHtmlBody();
+            }
+        };
+    }
+
+    private static Message convertMessage(MimeMessage message) throws IOException, MessagingException {
+        ByteArrayOutputStream rawMessage = new ByteArrayOutputStream();
+        message.writeTo(rawMessage);
+        return Message.Builder
+                .of()
+                .use(MimeConfig.PERMISSIVE)
+                .parse(new ByteArrayInputStream(rawMessage.toByteArray()))
+                .build();
+    }
+
+    private static Optional<HeadersDto> fetchHeaders(List<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.HEADERS)) {
+            return Optional.empty();
+        }
+
+        try {
+            return Optional.ofNullable(mail.getMessage())
+                    .map(Throwing.function((MimeMessage message) -> {
+                        HeadersDto headers = new HeadersDto();
+                        Collections
+                            .list(message.getAllHeaders())
+                            .forEach((header) -> headers.add(header.getName(), header.getValue()));
+                        return headers;
+                    }).sneakyThrow());
+        } catch (MessagingException e) {
+            throw new InnaccessibleFieldException(AdditionalField.HEADERS, e);
+        }
+    }
+
+    private static Optional<Map<String, HeadersDto>> fetchPerRecipientsHeaders(List<AdditionalField> additionalFields, Mail mail) {
+        if (!additionalFields.contains(AdditionalField.PER_RECIPIENTS_HEADERS)) {
+            return Optional.empty();
+        }
+        Map<String, HeadersDto> headers = new HashMap<>();
+        PerRecipientHeaders specificHeaders = mail.getPerRecipientSpecificHeaders();
+        Multimap<MailAddress, PerRecipientHeaders.Header> headersByRecipient = specificHeaders.getHeadersByRecipient();
+
+        for (MailAddress address : headersByRecipient.keySet()) {
+            headers.put(address.asString(), fetchPerRecipientHeader(headersByRecipient, address));
+        }
+
+        return Optional.of(headers);
+    }
+
+    private static HeadersDto fetchPerRecipientHeader(
+            Multimap<MailAddress, PerRecipientHeaders.Header> headersByRecipient,
+            MailAddress address) {
+        HeadersDto header = new HeadersDto();
+        headersByRecipient.get(address).forEach((rawHeader) -> header.add(rawHeader.getName(), rawHeader.getValue()));
+        return header;
+    }
+
+    private static Optional<Map<String, String>> fetchAttributes(List<AdditionalField> additionalFields, Mail mail) {
+        if (!additionalFields.contains(AdditionalField.ATTRIBUTES)) {
+            return Optional.empty();
+        }
+
+        return Optional.of(Iterators.toStream(mail.getAttributeNames())
```

Comment:
Ident should be space

Reply:
It's already space-indented, so I'm a bit lost

Reply:
Should be 4 space, not more.

Sorry I was numlock and did not saw it.